<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<meta http-equiv="x-ua-compatible" content="IE=9" >

<title>Using PLINK to permute GWAS data</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: rgb(88, 72, 246)
   }

   pre .number {
     color: rgb(0, 0, 205);
   }

   pre .comment {
     color: rgb(76, 136, 107);
   }

   pre .keyword {
     color: rgb(0, 0, 255);
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: rgb(3, 106, 7);
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>


<!-- MathJax scripts -->
<script type="text/javascript" src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>



</head>

<body>
<h1>Using PLINK to permute GWAS data</h1>

<p><a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/">PLINK</a> is a platform for analysing GWAS data.  Lots of features, etc.</p>

<p>Ineresting fact.  There is a &ldquo;<a href="https://www.cog-genomics.org/plink2">fork</a>&rdquo; of PLINK that is a collab b/w the original authors, BGI, and some other folks.  It appears to be geared more towards nextgen-based GWAS. We&#39;ll stick with 1.07 for now&hellip;</p>

<p>PLINK 1.9 is available as of May 16, 2014.  Claims to be faster.  Will investigate soon.</p>

<p>One possible use of PLINK would be to implement the ESM test using PLINK to do all the heavy lifting.  This document is a set of notes on the good and bad of PLINK, with special emphasis on permutation output vis-a-vis best practices on high-performance computing clusters.</p>

<h1>Compilation Notes</h1>

<p>I have installed v1.07 from source on my iMac (OS X Mavericks, clang compilers).  It needed modification to several source files to get rid of compiler errors.  Also, the &ldquo;SYS&rdquo; variable needs to be set in the Makefile.  Instructions are provided in the Makefile.  Default is for Unix, and I had to set it to &ldquo;MAC&rdquo;.</p>

<h1>Test data</h1>

<p>I downloaded &ldquo;example.zip&rdquo; from <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/res.shtml#teach">here</a>.  It contains the following stuff:</p>

<pre><code class="sh">ls -lhrt example/
total 196120
-rw-r--r--@ 1 krthornt  staff   6.7M May 24  2008 wgas1.map
-rw-r--r--@ 1 krthornt  staff   1.6K May 24  2008 pop.cov
-rw-r--r--@ 1 krthornt  staff   1.7M May 24  2008 gPLINK.jar
-rw-r--r--@ 1 krthornt  staff   509B May 24  2008 extra.map
-rw-r--r--@ 1 krthornt  staff   5.0M May 24  2008 Haploview.jar
-rw-r--r--@ 1 krthornt  staff    79M May 24  2008 wgas1.ped
-rw-r--r--@ 1 krthornt  staff   8.1K May 26  2008 extra.ped
-rw-r--r--@ 1 krthornt  staff   3.8M May 26  2008 plink.exe
-rw-r--r--@ 1 krthornt  staff   2.7K Jul 11  2008 command-list.txt
</code></pre>

<p>The file &ldquo;wgas1.ped&rdquo; is the main input file.  The PED format is described <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml#ped">here</a>.</p>

<p>Each PED file must have a corresponding MAP file, which is info about each SNP.  Details are <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml#map">here</a>.</p>

<h1>PLINK formatting issues</h1>

<p>There are several issues for how to write the data files that we need to address:</p>

<ol>
<li>For case/control data, 1 = control/unaffected, 2 = case/affected, 0 = missing.</li>
<li>We probably do not want to have sex identifiers, etc., in the PED files.</li>
<li>MAP files allow you to put in a negative number for the mutation position, indicating an excluded SNP.  This suggests the possibility of pre-filtering (based on HWE p-values, minimum allele counts, etc.), making a new MAP file representing the filtering, and then permuting.  However, the documentation says that this only works for plain-text input, not the binary-format input (which is desirable as it can be read into memory faster).  That may not matter&ndash;maybe the plain text is ok?</li>
</ol>

<h1>PLINK usage questions/comments</h1>

<p>Questions:</p>

<ol>
<li>Does PLINK have a way to subset the data?  For example, can we take large file and break it into chunks of \( M \) markers?  it <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/dataman.shtml#tabset">this</a> relevant?  Upon further poking around, <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/dataman.shtml#extract">this</a> is probably what we are looking for.  PLINK allows us to get all SNPs in a window, meaning we can break the genome up into windows of a constant size, and permute the SNPs w/in that window.</li>
</ol>

<p>Comments:</p>

<ol>
<li>It <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/thresh.shtml#hwd">looks like</a> we can exclude SNPs based on HWE in controls &ldquo;on the fly&rdquo;, which is a huge simplification.</li>
<li>It is critical to keep the following in mind: we wish to permute SNPs in windows sizes that are chosen for computational convenience.  However, we will slide windows of <strong>a different size</strong> over the data in order to calculate ESM.  This means that the permutations <strong>must be held constant, genome-wide</strong>, a la <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1206241/pdf/ge1383963.pdf">Churchill and Doerge</a>.  If we fail to do this, then windows used in the ESM calculations will be mixtures of different underling permutation samples.  In practice, this means that all perms from the same chromosome must use the <strong>same random number seed</strong>.</li>
</ol>

<h1>Basic usage</h1>

<p>Maybe a useful <a href="http://128.40.230.201/stupkalab/downloads/2010.02.17_Lescai_TutorialPLINK.pdf">reference</a>?</p>

<p>Vanilla association test.  The output is a \( \chi^2 \) statistic for each marker:</p>

<pre><code class="sh">#!sh

#Basic association test, HWE test, and MAF counting, all in 1 command.  That is pretty cool.
#The --counts option replaces allele frequencies with counts in the output
../plink-1.07-src/plink --file ../example/wgas1 --assoc --counts --hardy --freq
</code></pre>

<h1>Generating permutations of the data</h1>

<p>PLINK can generate permutations of the data, and output either &ldquo;empirical&rdquo; p-values or the permuted values of all test statistics.</p>

<p>Several examples:</p>

<pre><code class="sh">#!sh

#Exploring permutations.

#Generate 100 perms of the example data, with association statistics written out
#This command simply reports two types of permutation p-value for the association statistic
#../plink-1.07-src/plink --file ../example/wgas1 --assoc --mperm 100 --seed 101

#Let&#39;s save ALL the perms.
#The output from this is UGLY.  For K snps:
#perm number permstat1 permstat2 ... permstat_k
#The command below is the preferred input for ESM, as the chi-squared is rapid to calculate.  
#Although the p-value of ANY single-marker test can be used as the ESM input, 
#other tests can be 10-100x slower to compute PER MARKER, which is really, really bad when you are permuting
../plink-1.07-src/plink --file ../example/wgas1 --assoc --mperm 10000 --seed 101 --mperm-save-all

#We can do it w/logistic test.  This is SLOW, as expected
../plink-1.07-src/plink --file ../example/wgas1 --logistic --mperm 100 --seed 101 --mperm-save-all

#We can do it w/Fisher&#39;s Exact test.  Again, slow
../plink-1.07-src/plink --file ../example/wgas1 --fisher --mperm 100 --seed 101 --mperm-save-all
</code></pre>

<p>On my iMac, it took 37 minutes and 50 seconds to do the \( 10^4 \) permutations of the example data, which consists of 90 individuals and over \( 2 \times 10^5 \) SNPS:</p>

<pre><code class="sh">wc example/wgas1.ped 
      90 41165460 82332090 example/wgas1.ped
wc testing/plink.frq.count 
  228695 1600865 10977360 testing/plink.frq.count
</code></pre>

<p>That timing isn&#39;t so bad, but it would be useful to test how long it takes to do, say, \( 3 \times 10^6 \) permutations of \( 3\times 10^3 \) individuals for 50 or 100 SNPs.  Presumably PLINK is smart enough that the \( \chi^2 \) calculation is the slow part.  In other words, I&#39;m hoping that it is permuting indexes representing each individual rather than actually moving all the genotype data around.</p>

<p>The output file is 17GB, gzipping down to 5.3GB, and the gzipping took a LONG time.</p>

<p>Some issues immediately become apparent:</p>

<ol>
<li>PLINK writes output to the screen.  In a GE environment, we&#39;d need to redirect stdout and stderr to /dev/null, otherwise we&#39;d crush the file system with the .e/.o files.</li>
<li>PLINK is not buffering output internally.  Rather, it continuously writes to a file called &ldquo;plink.mperm.dump.all&rdquo;, which is plain-text.  This approach has several drawbacks.  First, it is not clear that a file system like gluster could handle this.<br/></li>
<li>PLINK has a set of default output file <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/reference.shtml#output">names</a>.  The default behavior is to name each output file plink.suffix, where suffix is a function of the arguments you are using.  The prefix may be changes via the &ndash;out option (see example somewhere down below).</li>
</ol>

<p>The reason why these issues are important is that we will need to split up the data into chunks of \( M \) markers, and generate a large number of permutations per chunks.  Because we cannot control the output name, we may need to make a subdirectory for each chunk.  Further, the uncompressed files are a problem in terms of space, but we can easily gzip after each run.</p>

<p>Some solutions for dealing w/the output include:</p>

<ol>
<li>gzipping immediately after the perms are generated.  This is a no-brainer, but time-intensive.</li>
<li>adding the output files to a growing tar-format archive.  This requires file-locking at the shell level, which I have no direct experience with.  In principle, these tar archives are &ldquo;mountable&rdquo; as if they were their own partition.  Again, no experience, and may have to ask Harry.</li>
<li>Can we use mkfifo to simultaneously permute and gzip?</li>
</ol>

<p>Yes, the named pipes do seem to work, which is great:</p>

<pre><code class="sh">#!sh

#Exploring permutations using named pipes

mkfifo plink.mperm.dump.all
../plink-1.07-src/plink --file ../example/wgas1 --assoc --mperm 100 --seed 101 --mperm-save-all &amp;
cat plink.mperm.dump.all | gzip &gt; plink.mperm.dump.all.gz
rm -f plink.mperm.dump.all
</code></pre>

<p>No idea what the effect of the named pipe technique would be on gluster load, etc.</p>

<p>Anyhow, let&#39;s redo the \( 10^4 \) perms of the test data using this technique:</p>

<pre><code class="sh">#!sh

mkfifo plink.mperm.dump.all
../plink-1.07-src/plink --file ../example/wgas1 --assoc --mperm 10000 --seed 101 --mperm-save-all &amp;
cat plink.mperm.dump.all | gzip &gt; plink.mperm.dump.all.gz
rm -f plink.mperm.dump.all
</code></pre>

<p>Executing the script gave the following results:</p>

<pre><code class="sh">time sh permutation2.sh

real  46m10.570s
user    19m55.896s
sys 0m38.645s
</code></pre>

<p>So, that&#39;s a 20% performance hit, but the benefit is that we skip the generation of one very large file.</p>

<h1>Permuting subsets of the data</h1>

<p>Let&#39;s try the following:</p>

<ol>
<li> Permute just a subset of the test data</li>
<li> Apply a HWE filter.  The filter will be liberal, so that SNPs are certain to get tossed.</li>
<li> Manipulate the output file names</li>
</ol>

<pre><code class="sh">#!sh

OFILEBASE=subset
OFILE=$OFILEBASE.mperm.dump.all
mkfifo $OFILE
#The arguments are nicely self-explanatory...
../plink-1.07-src/plink --file ../example/wgas1 --assoc --mperm 10000 --seed 101 --mperm-save-all --chr 2 --from-kb 5000 --to-kb 10000 --hwe 0.05 --out $OFILEBASE &amp;
cat $OFILE | gzip &gt; $OFILE.gz
rm -f $OFILE
</code></pre>

<p>Some of the info printed to stdout is encouraging, suggesting that SNPs are excluded based on rejected HWE in controls only (which is what the documentation says, so yay for that):</p>

<pre><code class="sh">5 markers to be excluded based on HWE test ( p &lt;= 0.05 )
  11 markers failed HWE test in cases
    5 markers failed HWE test in controls
</code></pre>

<p>Additional notes:</p>

<ol>
<li>If you input &ndash;to-kb values that are greater than the length of the chromosome, the program handles it fine, and simply does the requested operation up until the last SNP.</li>
<li><p>If your &ndash;from-kb is greater than the position of the last marker, PLINK just prints an error message to stderr</p>

<h1>Trying to tie it all together</h1></li>
</ol>

<p>We have figured out how to:</p>

<ol>
<li> Calculate $\chi<sup>2$</sup> for every marker</li>
<li> Subset the data based on physical ranges</li>
<li> Obtain the permutation distribution of the $\chi<sup>2$</sup> statistic for the subsetted data</li>
<li> Use named pipes to avoid writing the very large permutation files.</li>
<li> How to control the prefix of the output file names</li>
<li> If you pick a window where there are no markers, the program exist gracefully.</li>
</ol>

<p>When the run is successful, the output looks like:</p>

<pre><code class="sh">   ls -lhrt
total 128
-rw-r--r--+ 1 krthornt  staff   324B May 15 13:08 permutation2.sh
-rw-r--r--+ 1 krthornt  staff    46K May 15 13:09 subset.mperm.dump.all.gz
-rw-r--r--+ 1 krthornt  staff   2.0K May 15 13:09 subset.log
-rw-r--r--+ 1 krthornt  staff   264B May 15 13:09 subset.assoc.mperm
-rw-r--r--+ 1 krthornt  staff   576B May 15 13:09 subset.assoc
</code></pre>

<p>The files are:</p>

<ol>
<li>subset.log: a repeat of what is written to stdout.  This can be deleted, or they can all be archived for posterity&#39;s sake.</li>
<li>subset.assoc.mperm: the permutation p-value of the association test statistics.  We do not need this&ndash;delete</li>
<li>subset.assoc: the output of the association test.  This file is needed, as it contains the names of the SNPs that were analyzed, and is critical info, allowing us to read the MAP file back in and assign chromosome, position, etc.</li>
<li>subset.mperm.dump.all.gz: for each permuted data set, you get the permutation ID number, and then the permuted association test statistic for each marker, in the order that they appear in subset.assoc.  (That last statement is an assumption&ndash;not verified!)</li>
</ol>

<p>I think that the situation for actually doing the perms is simple:</p>

<ol>
<li> Choose a window size corresponding to a number of SNPs that is modest on average (50? 100?).</li>
<li> The output file name prefix can simply be chrC.start.stop, where C is the chromosome label (1-22, as the data are only autosomal), and the start/stop are the values passed to PLINK&#39;s &ndash;from-kb and &ndash;to-kb arguments.</li>
</ol>

<p>The appeal is that we need very few tools (see next section).  The caveat is whether or not HPC can handle the I/O load.  That&#39;s Harry&#39;s problem&hellip;</p>

<h1>So what do we need?</h1>

<ol>
<li> A program to convert WTCCC data into PLINK format</li>
<li> A program to merge the permutation output from PLINK back into a large file.  Given the sizes, I think we need to suck it up and write directly to gzipped files using the low-level <a href="http://zlib.net/">zlib</a> library, which allows appending, seeking, etc.  This file needs to be &ldquo;sensible&rdquo;, containing the SNP names from the WTCCC input files, so that they can be mapped directly to genes, etc., using current annotations.  Not doing this the first time through was a big impediment to downstream analysis.   In practice, this step merges the .assoc file with the .mperm.dump.all.gz file with the MAP file for the appropriate chromosome.</li>
<li> A program to perform the ESM test on the output of #2</li>
<li> We need to deal with LD between SNPS.  Not sure if this is a custom program to write, or if PLINK can do something automatically.</li>
</ol>

<p>Outside-the-box ideas:</p>

<ol>
<li> Instead of piping the named pipe to gzip, could we simply buffer the data internally and then write it to a tar file?  Problem: large data = risk of node crashing.</li>
</ol>

<h1>Dealing with linkage disequilibrium (LD)</h1>

<p>PLINK has <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/ld.shtml">functions</a> for calculating LD.  They are easy to use.</p>

<p>To calculate pairwise correlation coefficients ( $r<sup>2$</sup> ) for <strong>ALL</strong> pairs of markers:</p>

<pre><code class="sh">../plink-1.07-src/plink --file ../example/wgas1 --r2
</code></pre>

<p>For just a single chromosome, outputting to chr2.ld and removing markers with significant HWE deviations in controls:</p>

<pre><code class="sh">../plink-1.07-src/plink --file ../example/wgas1 --r2 --chr 2 --out chr2 --hwe 0.05
</code></pre>

<p>Sneakily, PLNK has some defaults that we need to over-write.  The following command calculates LD for all pairs of SNPs within 1000 SNPs of each other, within the same 10 megabase window (10,000kb), and only reports LD statistics &gt; 0.8:</p>

<pre><code class="sh">../plink-1.07-src/plink --file ../example/wgas1 --r2 0.8 --chr 2 --out chr2 --hwe 0.05 --ld-window 1000 --ld-window-kb 10000 --ld-window-r2 0.8
</code></pre>

<p>So, we just need to run that per chromosome once, gzip up the output, and store it.</p>

<h1>How much faster is the binary input?</h1>

<p>PLINK lets you convert the PED/MAP files from plain text into native <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml#bed">binary format</a>.</p>

<p>Converting results in <strong>much</strong> smaller files:</p>

<pre><code class="sh">#The command is easy
../plink-1.07-src/plink --file ../example/wgas1 --make-bed

ls -lhrt
total 25384
-rw-r--r--+ 1 krthornt  staff   281B May 16 15:25 btest.sh
-rw-r--r--+ 1 krthornt  staff   2.1K May 16 15:25 plink.fam
-rw-r--r--+ 1 krthornt  staff   1.7K May 16 15:25 plink.log
-rw-r--r--+ 1 krthornt  staff   7.4M May 16 15:25 plink.bim
-rw-r--r--+ 1 krthornt  staff   5.0M May 16 15:25 plink.bed

#Compare to old file sizes:
thornton01:binary_test krthornt$ ls -lhrt ../example/wgas1.*
-rw-r--r--+ 1 krthornt  staff   6.7M May 24  2008 ../example/wgas1.map
-rw-r--r--+ 1 krthornt  staff    79M May 24  2008 ../example/wgas1.ped
</code></pre>

<p>The binary input is also way faster to read in.</p>

<p>OK, so if the binary format is smaller and faster, it seems to be a win. But, it gets rid of our ability to multiply marker positions by -1 to exclude them.  It <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/dataman.shtml#exclude">looks like</a> we can provide SNP exclusion lists, and that this can be done at the stage where binary files are generated.</p>

<p>It is also possible to <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/dataman.shtml#remove">remove individuals</a> using a file of individuals to exclude.  Again, this can be done when making binary files.</p>

<h1>Random notes about PLINK</h1>

<p>Various insights copy/pasted from their website:</p>

<ol>
<li>If an input file is compressed (gzip compression) and ends in the .gz extension, PLINK will automatically decompress it (if compiled with ZLIB support).  This appears to only apply to the <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/metaanal.shtml">meta analysis module</a>.  Lame.</li>
</ol>

<h1>How permutation works in PLINK:</h1>

<p>This section corresponds to PLINK 1.07.</p>

<p>From helper.cpp,</p>

<pre><code class="cpp">void permute(vector&lt;long int&gt; &amp;a)
{
    // generate random permutation of 0..n-1
    // where n is a.size();

    const long int n = a.size( );

    for( long int i = 0; i &lt; n; i++ )
        a[ i ] = i;

    for( long int j = 1; j &lt; n; j++ )
    {
        long int pos = CRandom::rand(j+1);
        long int tmp = a[ j ];

        a[ j ] = a[ pos ];
        a[ pos ] = tmp;
    }
}
</code></pre>

<p>The above is called in perm.cpp as follows</p>

<pre><code class="cpp">void Perm::permuteInCluster()
{

  // Store remapped IDs
  vector&lt;vector&lt; long int&gt; &gt; i(ns);

  // Permute phenotypes, within cluster
  for (int k=0; k&lt;ns; k++)
    {
      vector&lt;long int&gt; p(s[k].size());
      permute(p);
      i[k]=p;
    }

  //////////////////////////
  // Post-permutation:
  // Iterate over clusters { s[][] }
  // i[][] holds the permuted codes
  // s[][] points to individuals (non-missing)

  // Genotype =           sample[s[j][k]];      
  // Matching phenotype = sample[s[j][(int)i[j][k]]];       

  // Create pheno[] with label-swapped codes
  for (int j=0; j&lt;s.size(); j++)
    for (int k=0; k&lt;s[j].size(); k++)
      P.sample[s[j][k]]-&gt;pperson = P.sample[s[j][(int)i[j][k]]];

}
</code></pre>

<p>It is pretty easy to see that permutation is swapping phenotypes amongst individuals w/in the same <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml#clst">cluster</a>.  (By default, there is one cluster in an analysis.)</p>

<p>Finally, PLINK sets the seed once and only once, in plink.cpp:</p>

<pre><code class="cpp">  if ( par::random_seed == 0 )
    CRandom::srand(time(0));
  else
    CRandom::srand( par::random_seed );
</code></pre>

<p>All of the above is important because it means the following:</p>

<ol>
<li> We can use 1 seed for each analysis.  In other words, HT gets 1 seed, BD another, etc.  Further, each subset of SNPs within an analysis uses <strong>the same seed</strong>, and therefore the permutation order is held constant genome-wide, because the individuals used in the analysis will be held constant genome-wide (e.g., the only exclusions of individuals will be based on the exclusion lists that come with the WTCCC data).</li>
</ol>

</body>

</html>

